hoodie-client/src/main/java/com/uber/hoodie/io/HoodieCommitArchiveLog.java

```diff
@@ -58,132 +70,198 @@ public HoodieCommitArchiveLog(HoodieWriteConfig config, FileSystem fs) {
-        return success & deleteOtherInstants();
     }
 
-    private boolean deleteOtherInstants() {
-        // Delete clean and rollback files
-        List<HoodieInstant> toDelete = getInstantsToDelete().collect(Collectors.toList());
-        if(!toDelete.isEmpty()) {
-            log.info("Deleting actions " + toDelete);
-            return deleteInstants(toDelete);
+    private void close() {
+        try {
+            if(this.writer != null) {
+                this.writer.close();
+            }
+        } catch(IOException e) {
+            throw new HoodieException("Unable to close HoodieLogFormat writer", e);
         }
-        return true;
     }
 
-    private Stream<HoodieInstant> getInstantsToDelete() {
-
-        int maxCommitsToKeep = config.getMaxCommitsToKeep();
-        int minCommitsToKeep = config.getMinCommitsToKeep();
-
-        HoodieTable table = HoodieTable.getHoodieTable(new HoodieTableMetaClient(fs, config.getBasePath(), true), config);
-        HoodieTimeline cleanTimeline = table.getActiveTimeline().getTimelineOfActions(Sets.newHashSet(HoodieTimeline.CLEAN_ACTION));
-        if (!cleanTimeline.empty() && cleanTimeline.countInstants() > maxCommitsToKeep) {
-            // Actually do the commits
-            return cleanTimeline.getInstants().limit(cleanTimeline.countInstants() - minCommitsToKeep);
+    /**
+     * Check if commits need to be archived. If yes, archive commits.
+     */
+    public boolean archiveIfRequired() {
+        try {
+            List<HoodieInstant> instantsToArchive = getInstantsToArchive().collect(Collectors.toList());
+            boolean success = true;
+            if (instantsToArchive.iterator().hasNext()) {
+                this.writer = openWriter();
+                log.info("Archiving instants " + instantsToArchive);
+                archive(instantsToArchive);
+                success = deleteArchivedInstants(instantsToArchive);
+            } else {
+                log.info("No Instants to archive");
+            }
+            return success;
+        } finally {
+            close();
         }
-        return Stream.empty();
     }
 
-    private Stream<HoodieInstant> getCommitsToArchive() {
+    private Stream<HoodieInstant> getInstantsToArchive() {
 
         int maxCommitsToKeep = config.getMaxCommitsToKeep();
         int minCommitsToKeep = config.getMinCommitsToKeep();
 
         HoodieTable table = HoodieTable.getHoodieTable(new HoodieTableMetaClient(fs, config.getBasePath(), true), config);
+
+        HoodieTimeline cleanAndRollbackTimeline = table.getActiveTimeline().getTimelineOfActions(Sets.newHashSet(HoodieTimeline.CLEAN_ACTION,
+                HoodieTimeline.ROLLBACK_ACTION));
+        Stream<HoodieInstant> instants = cleanAndRollbackTimeline.getInstants()
+            .collect(Collectors.groupingBy(s -> s.getAction()))
+            .entrySet()
+            .stream()
+            .map(i -> {
+                if (i.getValue().size() > maxCommitsToKeep) {
```

Comment:
add a comment here to understand that we are keeping `maxCommitsToKeep` per action..